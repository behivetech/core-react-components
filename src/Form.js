// Vendor Libs
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import {
    debounce,
    isEmpty,
    isFunction,
    noop,
    transform,
} from 'lodash';
import * as validators from './lib-app/validators';

const DEFAULT_CONTEXT = {
    getFieldError: noop,
    getFieldValue: noop,
    setFieldValidators: noop,
    setFieldValue: noop,
    setFieldValueDebounced: noop,
    shouldDisable: noop,
};

export const FormContext = React.createContext(DEFAULT_CONTEXT);

export default class Form extends Component {
    constructor(props) {
        super(props);

        this.getFieldValue = (fieldName) => {
            return this.state.fieldValues[fieldName];
        };

        this.getFieldError = (fieldName) => {
            const {fieldErrors, initialSubmit} = this.state;

            return (initialSubmit)
                ? fieldErrors[fieldName]
                : undefined;
        };

        this.setFieldValue = (fieldName, value) => {
            if (fieldName && value !== this.getFieldValue(fieldName)) {
                this.setState({
                    fieldValues: {
                        ...this.state.fieldValues,
                        [fieldName]: value,
                    },
                }, () => this.validateField(fieldName));
            }
        };

        this.setFieldValidators = (fieldName, validate) => {
            if (fieldName && validate) {
                this.setState({
                    fieldValidators: {
                        ...this.state.fieldValidators,
                        [fieldName]: validate,
                    },
                }, () => this.validateField(fieldName));
            }
        };

        this.shouldDisable = () => {
            return (this.state.submitting);
        };

        this.setFieldValueDebounced = debounce(this.setFieldValue, 200, {trailing: true});

        this.state = {
            context: {
                getFieldError: this.getFieldError,
                getFieldValue: this.getFieldValue,
                setFieldValidators: this.setFieldValidators,
                setFieldValue: this.setFieldValue,
                setFieldValueDebounced: this.setFieldValueDebounced,
                shouldDisable: this.shouldDisable,
            },
            fieldErrors: {},
            fieldValues: {},
            fieldValidators: {},
            initialSubmit: false,
            submitting: false,
        };
    }

    componentWillMount() {
        const {fieldValues} = this.props;

        if (fieldValues) {
            this.setState({fieldValues});
        }
    }

    componentWillReceiveProps({submitted: nextSubmitted}) {
        if (!this.props.submitted && nextSubmitted) {
            this.setState({submitting: false});
            // Do something once form has been submitted
        }
    }

    componentWillUpdate(nextProps, nextState) {
        const {submitting} = this.state;
        const {submitting: nextSubmitting} = nextState;
        const {onSubmit: nextOnSubmit} = nextProps;

        if (
            !submitting &&
            nextSubmitting &&
            !this.hasErrors(nextProps, nextState)
        ) {
            nextOnSubmit();
            this.setState({initialSubmit: false});
        }
    }

    setFieldError(fieldName, error) {
        const currentFieldError = this.state.fieldErrors[fieldName];

        if (
            fieldName &&
            (
                (error && error !== currentFieldError) ||
                (!error && currentFieldError)
            )
        ) {
            this.setState({
                fieldErrors: {
                    ...this.state.fieldErrors,
                    [fieldName]: error,
                },
            });
        }
    }

    validateField(fieldName, forceValidate) {
        const {fieldValues, fieldValidators} = this.state;
        const fieldValidatorsArray = fieldValidators[fieldName];
        let fieldError = null;

        if (fieldValidatorsArray && fieldValidatorsArray.length) {
            const fieldValue = this.getFieldValue(fieldName);

            // Using transform to iterate through the validators until an error is found.
            // Only sending one error at a time to the form field to prevent a mass
            // amount of errors showing up. This is also a much more efficient approach
            // so all validators aren't iterated through every time the field needs to
            // be validated.
            transform(fieldValidatorsArray, (errors, validator) => {
                if (isFunction(validator)) {
                    fieldError = validator(fieldValue, fieldValues);
                } else {
                    const importedValidator = validators[validator]; // eslint-disable-line

                    if (importedValidator && isFunction(importedValidator)) {
                        fieldError = importedValidator(fieldValue, fieldValues);
                    } else {
                        fieldError = `${validator} is not a valid validator.`;
                    }
                }

                return !fieldError;
            });

        }

        this.setFieldError(fieldName, fieldError);
    }

    hasErrors(props = this.props, state = this.state) {
        return (!isEmpty(state.fieldErrors) || !!props.formError);
    }

    handleSubmit = (event) => {
        event.preventDefault();
        this.setState({initialSubmit: true, submitting: true});
    }

    render() {
        const {children, className, name} = this.props;

        return (
            <FormContext.Provider value={this.state.context}>
                <form className={classnames('form', className)} name={name} onSubmit={this.handleSubmit}>
                    {children}
                </form>
            </FormContext.Provider>
        );
    }
}

Form.propTypes = {
    /** The children to go in the form */
    children: PropTypes.node,
    /** Used to add attitional styling from parent component */
    className: PropTypes.string,
    /** Any form type errors to help with it submitting and showing errors */
    formError: PropTypes.string,
    /**
        Values to populate the fields going by the same name you use
        as the name prop for the field and using the FormField component
    */
    fieldValues: PropTypes.objectOf(PropTypes.oneOfType([
        PropTypes.array,
        PropTypes.number,
        PropTypes.string,
    ])),
    /**
        Standard name you would give a form using html. Helps with react refs.
        It's ideal to keep these names unique per what's displayed on the current
        page because of the potential refs created.
        (refs are coming from a FormRefProvider component to have the ability to
        submit the form outside of the form).
    */
    // TODO: need to figure out the ref thing and disabling components that would
    // potentially submit this form from outside of the form. Probably pass the
    // values down through the value argument in the provider and callback functions.
    name: PropTypes.string.isRequired,
    /**
        Functionality to run when this Form component meets all requirements
        such as passsing validation tests.
    */
    onSubmit: PropTypes.func,
    /** Indicates if the form was submitted and a response came back */
    submitted: PropTypes.bool,
};

Form.defaultProps = {
    onSubmit: noop,
};
