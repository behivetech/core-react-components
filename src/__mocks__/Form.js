const context = {
    getFieldError: jest.fn(),
    getFieldValue: jest.fn(),
    setFieldValidators: jest.fn(),
    setFieldValue: jest.fn(),
    setFieldValueDebounced: jest.fn(),
    shouldDisable: jest.fn(),
};

export const FormContext = ({
    Consumer(props) {
        return props.children(context);
    },
});