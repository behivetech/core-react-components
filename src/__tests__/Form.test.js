import Form from "../Form";

describe('Form', () => {
    const getProps = (updateProps) => {
        return {
            className: 'mock-class',
            name: 'mockForm',
            ...updateProps,
        };
    };

    it('should render with the correct state and props', () => {
        const mockProps = {fieldValues: {mockField: 'mock value'}};
        const wrapper = mount(<Form {...getProps(mockProps)} />);
        const {context, fieldValues} = wrapper.state();

        expect(wrapper.debug()).toMatchSnapshot();
        expect(fieldValues).toEqual(mockProps.fieldValues);
        expect(context).toMatchSnapshot();
    });

    it('should set and unset a field error', () => {
        const wrapper = mount(<Form {...getProps()} />);
        const wrapperStateContext = wrapper.state().context;

        wrapperStateContext.setFieldValidators('mockField1', ['required']);
        expect(wrapper.state().fieldErrors.mockField1).toBe('This field cannot be empty');
        wrapperStateContext.setFieldValue('mockField1', 'mock value');
        expect(wrapper.state().fieldErrors.mockField1).toBe(undefined);
    });

    it('should call a validator function and set the first true error', () => {
        const wrapper = mount(<Form {...getProps()} />);
        const mockValidator = (error) => () => error;

        wrapper.state().context.setFieldValidators(
            'mockField',
            [mockValidator(), mockValidator('error 1'), mockValidator('error 2')]
        );

        expect(wrapper.state().fieldErrors.mockField).toEqual('error 1');
    });

    it('should set an error message when the type of validator does not match', () => {
        const wrapper = mount(<Form {...getProps()} />);

        wrapper.state().context.setFieldValidators(
            'mockField',
            ['badMockValidator']
        );

        expect(wrapper.state().fieldErrors.mockField).toEqual('badMockValidator is not a valid validator.');
    });

    it('should not set a field validator unless all arguments are correct', () => {
        const mockSetState = jest.fn();
        const wrapper = mount(<Form {...getProps()} />);
        const wrapperInstance = wrapper.instance();

        wrapperInstance.setState = mockSetState;
        wrapperInstance.setFieldValidators(null, 'mockValidator');
        expect(mockSetState).not.toHaveBeenCalled();
        wrapperInstance.setFieldValidators('mockField');
        expect(mockSetState).not.toHaveBeenCalled();

    });

    it('should set a field error only when arguments pass conditions', () => {
        const mockSetState = jest.fn();
        const wrapper = mount(<Form {...getProps()} />);
        const wrapperInstance = wrapper.instance();
        const wrapperSetState = wrapperInstance.setState;

        wrapperInstance.setState = mockSetState;
        wrapperInstance.setFieldError(null, 'mock error');
        expect(mockSetState).not.toHaveBeenCalled();
        wrapperInstance.setFieldError('mockField');
        expect(mockSetState).not.toHaveBeenCalled();
        wrapperInstance.setState = wrapperSetState;
        wrapper.setState({fieldErrors: {mockField: 'mock error'}});
        wrapperInstance.setState = mockSetState;
        wrapperInstance.setFieldError('mockField', 'mock error');
        expect(mockSetState).not.toHaveBeenCalled();
        wrapperInstance.setFieldError('mockField', 'mock error 1');
        expect(wrapperInstance.setState).toHaveBeenCalledWith({
            fieldErrors: {
                mockField: 'mock error 1',
            }},
        );
    });

    it('should only get a field error of the form has been initially submitted', () => {
        const wrapper = mount(<Form {...getProps()} />);
        const mockValidator = (error) => () => error;
        const wrapperState = wrapper.state();

        wrapperState.context.setFieldValidators(
            'mockField',
            [mockValidator('error 1')]
        );

        expect(wrapperState.context.getFieldError('mockField')).toBe(undefined);

        wrapper.setState({initialSubmit: true});

        expect(wrapperState.context.getFieldError('mockField')).toBe('error 1');
    });

    it('should or should not set a field value when conditions are met', () => {
        const mockProps = getProps({fieldValues: {mockField: 'mock value'}});
        const wrapper = mount(<Form {...mockProps} />);
        const wrapperStateContext = wrapper.state().context;
        const mockSetState = jest.fn();

        wrapper.instance().setState = mockSetState;
        wrapperStateContext.setFieldValue('mockField', 'mock value');
        expect(mockSetState).not.toHaveBeenCalled();
        wrapperStateContext.setFieldValue('mockField', 'mock value 1');
        expect(mockSetState).toHaveBeenCalledWith(
            {fieldValues: {mockField: 'mock value 1'}},
            expect.any(Function)
        );
    });

    it('should handle a submit correctly', () => {
        const mockEvent = {preventDefault: jest.fn()};
        const wrapper = mount(<Form {...getProps()} />);

        wrapper.instance().handleSubmit(mockEvent);
        expect(mockEvent.preventDefault).toHaveBeenCalled();
        expect(wrapper.state().initialSubmit).toBe(true);
    });
});
