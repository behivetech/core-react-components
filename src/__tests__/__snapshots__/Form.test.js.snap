// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render with the correct state and props 1`] = `
"<Form className=\\"mock-class\\" name=\\"mockForm\\" fieldValues={{...}} onSubmit={[Function: noop]}>
  <form className=\\"form mock-class\\" name=\\"mockForm\\" onSubmit={[Function]} />
</Form>"
`;

exports[`Form should render with the correct state and props 2`] = `
Object {
  "getFieldError": [Function],
  "getFieldValue": [Function],
  "setFieldValidators": [Function],
  "setFieldValue": [Function],
  "setFieldValueDebounced": [Function],
  "shouldDisable": [Function],
}
`;
